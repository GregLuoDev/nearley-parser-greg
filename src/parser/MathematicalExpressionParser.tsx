"use client";

import { useState } from "react";
import nearley from "nearley";
import grammar from "../math-grammar/math.js"; // compiled grammar
import { IAst } from "./type.js";

export function MathematicalExpressionParser() {
  const [value, setValue] = useState("");
  const [ast, setAst] = useState<IAst | null>(null);
  const result = ast?.value;
  const [error, setError] = useState("");

  const parser = new nearley.Parser(
    nearley.Grammar.fromCompiled(grammar as nearley.CompiledRules)
  );
  const errorMessage =
    "Parse failed.\nThis parser can handle both arithmetic operations and comparison operators, following the rules of operator precedence and proper evaluation.\nPlease check your input.";

  function parse() {
    try {
      parser.feed(value);

      const results = parser.results;
      if (results.length) {
        setAst(results[0]);
        setError("");
      } else {
        setError(errorMessage);
      }
    } catch (error: unknown) {
      let message = "Parse failed. ";
      if (error instanceof Error) {
        message += "Error.message:" + error.message;
      } else {
        message += JSON.stringify(error);
      }
      console.error(message);

      setError(errorMessage);
      setAst(null);
    }
  }

  function handleKeyDown(e: React.KeyboardEvent<HTMLInputElement>) {
    if (e.key === "Enter") {
      parse();
    }
  }

  return (
    <div className="flex flex-col bg-gray-100 p-6 min-h-[600px] max-w-xl w-full">
      <h1 className="text-2xl font-bold text-gray-700">
        Mathematical Expressions Parser
      </h1>

      <div className="flex flex-col sm:flex-row items-center gap-3 sm:gap-4 w-full max-w-xl mx-auto mt-6">
        <input
          type="text"
          value={value}
          placeholder="Type your mathematical expression"
          onChange={(e) => setValue(e.target.value)}
          className="flex-1 w-full sm:w-auto px-4 py-2 border border-gray-300 rounded-md shadow-sm focus:outline-none focus:ring-2 focus:ring-blue-500"
          data-testid="input"
          onKeyDown={handleKeyDown}
        />
        <button
          onClick={parse}
          className="px-5 py-2 bg-blue-600 text-white rounded-md hover:bg-blue-700 transition w-full sm:w-auto"
          data-testid="parser-button"
        >
          Parse
        </button>
      </div>

      {(result === true || result === false) && (
        <div className="mt-6 bg-green-100  border border-green-300 rounded-lg px-4 py-3 shadow">
          <span>Parsing Result:</span>{" "}
          <b
            className={`${result ? "text-green-800" : "text-red-800"}`}
            data-testid="result"
          >
            {result?.toString()}
          </b>
        </div>
      )}

      {ast && (
        <pre className="mt-6 bg-gray-100 rounded text-sm overflow-x-auto">
          <b>AST:</b>{" "}
          <div data-testid="ast">{JSON.stringify(ast, null, 4)}</div>
        </pre>
      )}

      {error && (
        <div className="mt-6 w-full bg-red-100 text-red-800 border border-red-300 rounded-lg p-4 shadow whitespace-pre-line">
          <b>Error:</b> <div data-testid="error">{error}</div>
        </div>
      )}
    </div>
  );
}
