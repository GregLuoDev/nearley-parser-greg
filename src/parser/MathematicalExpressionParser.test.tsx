import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { MathematicalExpressionParser } from "./MathematicalExpressionParser";

describe("MathematicalExpressionParser should render correctly", () => {
  beforeEach(() => {
    render(<MathematicalExpressionParser />);
  });

  test("should match snapshot", () => {
    const { container } = render(<MathematicalExpressionParser />);
    expect(container).toMatchSnapshot();
  });

  test("renders input box and parser button", () => {
    const input = screen.getByTestId("input");
    const button = screen.getByTestId("parser-button");

    expect(input).toBeInTheDocument();
    expect(button).toBeInTheDocument();
  });
});

describe("MathematicalExpressionParser shows result and AST for valid expressions without error", () => {
  beforeEach(() => {
    render(<MathematicalExpressionParser />);
  });

  test.each([
    ["1+2=3", true],
    [" 1+2=3 ", true],
    ["1 + 2 = 3", true],
    ["   1   +   2   = 3   ", true],
    ["6 = 10 / 2 + 1", true],
    ["12 + 3 != 4 / 2 + 5", true],
    ["2 + 3 * 2 = 10", false],
    ["2 * 3 + 4 != 10", false],
    [" 10 - 4 = 100 / 10 - 10 + 3 * 2 ", true],
    ["10 - 4 != 100 / 10 - 10 + 3 * 2 ", false],
    [" 10 - 4 != 100 / 10 - 10 + 3 * 2 +1", true],
    [" 2 * 3 + 4 = 20 - 5 -10 + 3 + 2 * 1 ", true],
    ["2 * 3 + 4 != 20 - 5 -10 + 3 + 2 * 1 ", false],
    [" 2 * 3 + 4 != 20 - 5 -10 + 3 + 2 * 1 +1", true],
    [" 8 / 2 = 2 * 2 ", true],
    ["8 / 2 != 2 * 2 ", false],
    [" 8 / 2 != 2 * 2 +1", true],
  ])("parses '%s' to be %s with result and AST.", async (expression, expectedResult) => {
    const input = screen.getByTestId("input");
    const button = screen.getByTestId("parser-button");

    fireEvent.change(input, { target: { value: expression } });
    fireEvent.click(button);

    const result = await screen.findByTestId("result");
    const ast = await screen.findByTestId("ast");

    expect(result).toBeInTheDocument();
    expect(result).toHaveTextContent(expectedResult.toString());
    expect(ast).toBeInTheDocument();

    const error = screen.queryByTestId("error");
    expect(error).not.toBeInTheDocument();
  });
});

describe("MathematicalExpressionParser shows error for invalid expressions without result and AST.", () => {
  beforeEach(() => {
    render(<MathematicalExpressionParser />);
  });

  test.each([
    [""],
    ["  "],
    ["123"],
    ["#1+2=3"],
    [" 1+(2=3 "],
    ["1 + 2 = 3*"],
    ["   1   +   2   =    "],
    ["^6 = 10 / 2 + 1"],
    ["12 + 3 != 4 / 2 + 5%"],
    ["2 + 3 * 2 = 10="],
    ["*2 * 3 + 4 != 10"],
    [" 10  4 = 100 / 10 - 10 + 3 * 2 "],
    [" 10 - 4 !3= 100 / 10 - 10 + 3 * 2 "],
    [" 10 - 4 != 100 / 10  10 + 3 * 2 +1"],
    [" 2 * 3 + 4 = 20 - 5 --10 + 3 + 2 * 1 "],
    [" 2 * 3 + 4 != 20 - 5 -10 + 3 + 2 * 1 * "],
    [" 2 * 3 + 4 != 20 - 5 -10 + 3 + 2 * 1 +1 $"],
    [" 8 / 2  2 * 2 "],
    [" 8  2 != 2 * 2 "],
    ["- 8 / 2 != 2 * 2 +1"],
  ])("parses '%s' failed with error message without result and AST", async (expression) => {
    const input = screen.getByTestId("input");
    const button = screen.getByTestId("parser-button");

    fireEvent.change(input, { target: { value: expression } });
    fireEvent.click(button);

    const error = await screen.findByTestId("error");

    expect(error).toBeInTheDocument();
    expect(error.textContent).toMatch(/Parse failed/i);

    const result = screen.queryByTestId("result");
    const ast = screen.queryByTestId("ast");
    expect(result).not.toBeInTheDocument();
    expect(ast).not.toBeInTheDocument();
  });
});
