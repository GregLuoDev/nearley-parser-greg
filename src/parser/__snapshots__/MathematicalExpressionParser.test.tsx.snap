// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`MathematicalExpressionParser should render correctly should match snapshot 1`] = `
<div>
  <h1
    class="text-2xl font-bold text-gray-700"
  >
    Mathematical Expressions Parser
  </h1>
  <div
    class="flex flex-col sm:flex-row items-center gap-3 sm:gap-4 w-full max-w-xl mx-auto mt-6"
  >
    <input
      class="flex-1 w-full sm:w-auto px-4 py-2 border border-gray-300 rounded-md shadow-sm focus:outline-none focus:ring-2 focus:ring-blue-500 text-black"
      data-testid="input"
      placeholder="Type your mathematical expression"
      type="text"
      value=""
    />
    <button
      class="px-5 py-2 bg-blue-600 text-white rounded-md hover:bg-blue-700 transition w-full sm:w-auto"
      data-testid="parser-button"
    >
      Parse
    </button>
  </div>
</div>
`;
